### ISSUE TITLE
tests are needlessly forced to run with 1 logical core (worker)

### ISSUE THREAD

**Issue by Apteryks** [JuliaLang/julia#43205]
Hello,

See: https://github.com/JuliaLang/julia/blob/master/Makefile#L541:

```
	@env JULIA_CPU_THREADS=1 $(MAKE) $(QUIET_MAKE) -C $(BUILDROOT)/test all JULIA_BUILD_MODE=$(JULIA_BUILD_MODE)
```

This makes the test suite to take much more time to execute; furthermore, the official CI uses multiple workers by manually invoking runtests.jl with the `ncores` argument, so there doesn't seem to be a reason to force 1 worker anymore (if there ever was).

Thank you!

**Comment by DilumAluthge** [comment]
This line is only relevant if you do `make testall1`.

Do `make testall` instead.

**Comment by Apteryks** [comment]
Ah, apologies for misreading that.  Still, using something like `make test -j24` causes the test suite to run only with 1 worker, e.g.:

```
exceptions                         (1) |        started at 2021-11-23T23:39:13.994
exceptions                         (1) |     1.87 |   1.26 | 67.2 |      30.02 |  7623.59
file                               (1) |        started at 2021-11-23T23:39:15.864
file                               (1) |     7.68 |   0.00 |  0.0 |     522.21 |  7623.59
read                               (1) |        started at 2021-11-23T23:39:23.545
read                               (1) |    19.41 |   0.71 |  3.7 |    2731.70 |  8003.43
version                            (1) |        started at 2021-11-23T23:39:42.956
version                            (1) |     0.81 |   0.00 |  0.0 |      58.98 |  8003.43
namedtuple                         (1) |        started at 2021-11-23T23:39:43.765
namedtuple                         (1) |     3.94 |   0.00 |  0.0 |     215.17 |  8003.43
mpfr                               (1) |        started at 2021-11-23T23:39:47.705
mpfr                               (1) |    11.99 |   0.37 |  3.1 |    1320.64 |  8003.43
broadcast                          (1) |        started at 2021-11-23T23:39:59.697
broadcast                          (1) |    72.92 |   1.79 |  2.5 |    5914.05 |  8003.43
complex                            (1) |        started at 2021-11-23T23:41:12.622
complex                            (1) |     8.99 |   1.17 | 13.0 |     807.41 |  8003.43
floatapprox                        (1) |        started at 2021-11-23T23:41:21.610
floatapprox                        (1) |     0.53 |   0.00 |  0.0 |      54.75 |  8003.43
reflection                         (1) |        started at 2021-11-23T23:41:22.138

reflection                         (1) |     6.28 |   0.00 |  0.0 |     540.67 |  8003.43
regex                              (1) |        started at 2021-11-23T23:41:28.423
regex                              (1) |     1.64 |   0.00 |  0.0 |     144.26 |  8003.43
float16                            (1) |        started at 2021-11-23T23:41:30.068
float16                            (1) |     1.89 |   0.00 |  0.0 |     159.42 |  8003.43
combinatorics                      (1) |        started at 2021-11-23T23:41:31.963
combinatorics                      (1) |     3.10 |   0.31 | 10.0 |     596.63 |  8240.51
sysinfo                            (1) |        started at 2021-11-23T23:41:35.060
sysinfo                            (1) |     0.11 |   0.00 |  0.0 |      15.76 |  8240.51
env                                (1) |        started at 2021-11-23T23:41:35.169
env                                (1) |     0.66 |   0.00 |  0.0 |      21.30 |  8240.51
rounding                           (1) |        started at 2021-11-23T23:41:35.828
rounding                           (1) |     4.92 |   0.00 |  0.0 |     440.08 |  8240.51
ranges                             (1) |        started at 2021-11-23T23:41:40.744
ranges                             (1) |    35.53 |   2.47 |  7.0 |    9558.68 |  8240.51
mod2pi                             (1) |        started at 2021-11-23T23:42:16.344
mod2pi                             (1) |     0.23 |   0.00 |  0.0 |      18.14 |  8240.51
euler                              (1) |        started at 2021-11-23T23:42:16.579
euler                              (1) |     1.46 |   0.25 | 17.4 |     100.04 |  8240.51
show                               (1) |        started at 2021-11-23T23:42:18.041
show                               (1) |    26.79 |   2.12 |  7.9 |    4061.38 |  8240.51
client                             (1) |        started at 2021-11-23T23:42:44.832
client                             (1) |     1.27 |   0.00 |  0.0 |      91.23 |  8240.51
```
I've tried setting JULIA_CPU_THREADS=24 but that didn't change a thing.  Any ideas?

**Comment by DilumAluthge** [comment]
When running `make test` or `make testall` for Julia, the `-j` flag to `make` does not serve any purpose.

By default, Julia should pick an appropriate number of workers if you run `make test` or `make testall`. If you want to control the number of workers, do e.g. `export JULIA_CPU_THREADS=32`.

Note that in order to use multiple test workers, we must have access to the loopback interface. What is the output of running the following command after you build Julia?
```
./julia -e 'Sockets.getipaddr()'
```

Also, what is the output of the following command?
```
./julia -e 'import InteractiveUtils; InteractiveUtils.versioninfo()'
```

**Comment by Apteryks** [comment]
Ah, that must be it; there is a private loopback network namespace in the build container, but:

```
$ ./julia -e 'Sockets.getipaddr()'
ERROR: UndefVarError: Sockets not defined
Stacktrace:
 [1] top-level scope
   @ none:1
```
and 
```
$ ./julia -e 'import InteractiveUtils; InteractiveUtils.versioninfo(; verbose = true)'  
Julia Version 1.6.3
Commit ae8452a9e0 (2021-09-23 17:34 UTC)
Platform Info:
  OS: Linux (x86_64-unknown-linux-gnu)
  uname: Linux 5.14.15-gnu #1 SMP 1 x86_64 unknown
  CPU: AMD Ryzen 9 3900X 12-Core Processor: 
                 speed         user         nice          sys         idle          irq
       #1-24  2200 MHz   31645684 s     438366 s    3329276 s  401819359 s          0 s
       
  Memory: 31.309677124023438 GB (20367.7578125 MB free)
  Uptime: 1.824802e6 sec
  Load Avg:  0.06  0.54  0.52
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.0 (ORCJIT, znver2)
Environment:
  PKG_CONFIG_PATH = /gnu/store/v168wafwaqdj7grq2b2i4syda3ivw9h9-profile/lib/pkgconfig
  GUIX_PYTHONPATH = /gnu/store/v168wafwaqdj7grq2b2i4syda3ivw9h9-profile/lib/python2.7/site-packages
  HOME = /home/$USER
  CPLUS_INCLUDE_PATH = /gnu/store/v168wafwaqdj7grq2b2i4syda3ivw9h9-profile/include/c++:/gnu/store/v168wafwaqdj7grq2b2i4syda3ivw9h9-profile/include
  LIBRARY_PATH = /gnu/store/v168wafwaqdj7grq2b2i4syda3ivw9h9-profile/lib
  GUIX_LOCPATH = /gnu/store/v168wafwaqdj7grq2b2i4syda3ivw9h9-profile/lib/locale
  PATH = /gnu/store/v168wafwaqdj7grq2b2i4syda3ivw9h9-profile/bin:/gnu/store/v168wafwaqdj7grq2b2i4syda3ivw9h9-profile/sbin
  C_INCLUDE_PATH = /gnu/store/v168wafwaqdj7grq2b2i4syda3ivw9h9-profile/include
  CPATH = /gnu/store/v168wafwaqdj7grq2b2i4syda3ivw9h9-profile/include
```

**Comment by Apteryks** [comment]
Oh, hum... That's embarrassing, but I get the same thing outside of the minimal build environment (where networking is fully functional).  So it seems our packaging of Julia on GNU Guix is broken in some way.  Would you know what may lead to this failure?  Perhaps the libuv library?  Thank you for your time.

**Comment by Apteryks** [comment]
I've tested that it works outside of the isolated container.  More debug outputs:

```
julia> Sockets.getipaddr("IPv4")
ERROR: MethodError: no method matching getipaddr(::String)
Closest candidates are:
  getipaddr() at /tmp/guix-build-julia-1.6.3.drv-0/julia-1.6.3/usr/share/julia/stdlib/v1.6/Sockets/src/addrinfo.jl:276
  getipaddr(!Matched::Type{T}) where T<:Sockets.IPAddr at /tmp/guix-build-julia-1.6.3.drv-0/julia-1.6.3/usr/share/julia/stdlib/v1.6/Sockets/src/addrinfo.jl:265
Stacktrace:
 [1] top-level scope
   @ none:1
```

**Comment by zimoun** [comment]
> I couldn't get the debugger going in the container (there is no `julia-debugger` package in Guix, it seems),

@Apteryks  `Debugger` would not help here, IMHO.  First install it using `Pkg`, then use it inside the container.

```
guix shell julia -- julia -e 'using Pkg; Pkg.add("Debugger")
guix shell -C julia -- julia -e 'using Sockets; using Debugger; @enter Sockets.getipaddr()'
```
It returns an irrelevant error (similarly from the REPL, i.e., `julia> @enter Sockets.getipaddr()`):

```
WARNING: failed to select UTF-8 encoding, using ASCII
ERROR: Debugger.jl needs to be run in a Julia REPL
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] RunDebugger(frame::JuliaInterpreter.Frame, repl::Nothing, terminal::Nothing; initial_continue::Bool)
   @ Debugger ~/.julia/packages/Debugger/I4w2y/src/repl.jl:15
 [3] RunDebugger (repeats 2 times)
   @ ~/.julia/packages/Debugger/I4w2y/src/repl.jl:13 [inlined]
 [4] top-level scope
   @ ~/.julia/packages/Debugger/I4w2y/src/Debugger.jl:127
```

**Comment by fredrikekre** [comment]
Duplicate of https://github.com/JuliaLang/julia/pull/40348

**Comment by zimoun** [comment]
@fredrikekre Thanks for the pointer.

I am sorry but I do not understand the current situation about running tests in parallel for isolated environments.  For instance, #40359 leads to the addition of the comment:

https://github.com/JuliaLang/julia/blob/master/test/runtests.jl#L94

But nothing is changed regarding the if-test about `net_on`; which is #40348  Does it mean it is not possible to run test in parallel for isolated environments?  Do we have to patch `runtests.jl`?

Thanks

**Comment by Apteryks** [comment]
@zimoun looks like it currently requires patching, yes: https://github.com/NixOS/nixpkgs/blob/c43789e7bb3768e0a084e4bf79366212d4a69f4b/pkgs/development/compilers/julia/patches/1.6-bin/0005-nix-Enable-parallel-unit-tests-for-sandbox.patch

**Comment by zimoun** [comment]
@fredrikekre @DilumAluthge Fixed by PR #43211 I guess.  Let me know.

Cheers,
simon
